crates/winch/src/visitor.rs

```diff
@@ -0,0 +1,713 @@
+	    @relaxed_simd I32x4RelaxedLaneselect => visit_i32x4_relaxed_laneselect
+	    @relaxed_simd I64x2RelaxedLaneselect => visit_i64x2_relaxed_laneselect
+	    @relaxed_simd F32x4RelaxedMin => visit_f32x4_relaxed_min
+	    @relaxed_simd F32x4RelaxedMax => visit_f32x4_relaxed_max
+	    @relaxed_simd F64x2RelaxedMin => visit_f64x2_relaxed_min
+	    @relaxed_simd F64x2RelaxedMax => visit_f64x2_relaxed_max
+	    @relaxed_simd I16x8RelaxedQ15mulrS => visit_i16x8_relaxed_q15mulr_s
+	    @relaxed_simd I16x8DotI8x16I7x16S => visit_i16x8_dot_i8x16_i7x16_s
+	    @relaxed_simd I32x4DotI8x16I7x16AddS => visit_i32x4_dot_i8x16_i7x16_add_s
+	    @relaxed_simd F32x4RelaxedDotBf16x8AddF32x4 => visit_f32x4_relaxed_dot_bf16x8_add_f32x4
+	}
+    };
+}
+
+/// A macro to define the corresponding methods for both supported and unsupported operators
+macro_rules! define {
+    ($( @$proposal:ident $op:ident $({ $($arg:ident: $argty:ty),*})? => ($visit:ident $emit:ident))*) => {
+        $(
+            fn $visit(&mut self, offset: usize $($(,$arg: $argty)*)?) -> Self::Output {
+		self.emitop(
+		    |validator| validator.$visit(offset $($(,$arg)*)?),
+		    |env| env.$emit($($($arg),*)?)
+		)
+            }
+        )*
+    };
+
+    ($( @$proposal:ident $op:ident $({ $($arg:ident: $argty:ty),*})? => $visit:ident )*) => {
+        $(
+            fn $visit(&mut self, _offset: usize $($(,$arg: $argty)*)?) -> Self::Output {
+		todo!(stringify!($op))
+            }
+        )*
+    };
+}
+
+impl<'c, 'a: 'c, M> CodeGen<'a, 'c, M>
+where
+    M: MacroAssembler,
+{
+    fn emitop<V, E>(&mut self, validate: V, emitter: E) -> Result<()>
+    where
+        V: FnOnce(&mut FuncValidator<ValidatorResources>) -> ValidationResult<()>,
+        E: FnOnce(&mut Self) -> Result<()>,
+    {
+        validate(&mut self.validator)
+            .map_err(|e| e.into())
+            .and(emitter(self))
+    }
+
+    fn emit_i32_add(&mut self) -> Result<()> {
+        let is_const = self
+            .context
+            .stack
+            .peek()
+            .expect("value at stack top")
+            .is_i32_const();
+
+        if is_const {
+            self.add_imm_i32();
+        } else {
+            self.add_i32();
+        }
+
+        Ok(())
+    }
+
+    fn add_imm_i32(&mut self) {
+        let val = self
+            .context
+            .stack
+            .pop_i32_const()
+            .expect("i32 constant at stack top");
+        let reg = self
+            .regalloc
+            .pop_to_reg(&mut self.context, OperandSize::S32);
+        self.context
+            .masm
+            .add(RegImm::imm(val), RegImm::reg(reg), OperandSize::S32);
+        self.context.stack.push(Val::reg(reg));
+    }
+
+    fn add_i32(&mut self) {
+        let src = self
+            .regalloc
+            .pop_to_reg(&mut self.context, OperandSize::S32);
+        let dst = self
+            .regalloc
+            .pop_to_reg(&mut self.context, OperandSize::S32);
+
+        self.context
```

Comment:
Ah, now I understand why two-arg `add` is kind of OK actually: Wasm `add` (and other binops) always consume their args, so it's fine to overwrite `dst`. That actually makes a lot of sense to me, so maybe it's fine to keep two-arg forms of the masm methods, unless it later hinders lowerings of e.g. SIMD ops or address computations or ... that use temps.